src/mscorlib/shared/System/IO/FileStreamCompletionSource.Win32.cs

```diff
@@ -217,6 +217,29 @@ private static void Cancel(object state)
                     }
                 }
             }
+
+            public static FileStreamCompletionSource Create(FileStream stream, int numBufferedBytesRead, ReadOnlyMemory<byte> memory)
+            {
+                if (MemoryMarshal.TryGetArray(memory, out ArraySegment<byte> arraySegment))
+                {
+                    byte[] buffer = arraySegment.Array;
+
+                    // If the memory passed in isn't the stream's internal buffer, then delegate to
+                    // memory.Retain instead of passing the buffer to AllocateNativeOverlapped. In 
+                    // some cases, this will result in less pinning (in case the underlying memory is backed
+                    // by pre-pinned buffers)
+                    if (!ReferenceEquals(buffer, stream._buffer))
+                    {
+                        // Nothing to pin here, fall through to the MemoryFileStreamCompletionSource path
+                    }
+                    else
+                    {
+                        return new FileStreamCompletionSource(stream, numBufferedBytesRead, buffer);
+                    }
+                }
+
+                return new MemoryFileStreamCompletionSource(stream, numBufferedBytesRead, memory);
```

Comment:
Nit: It's ok if you prefer it this way; if it were me, I'd have written it at least with the condition inverted, and maybe condensed to more like:
```C#
// If the memory passed in is the stream's internal buffer, we can use the base FileStreamCompletionSource,
// which has a PreAllocatedOverlapped with the memory already pinned.  Otherwise, we use the derived
// MemoryFileStreamCompletionSource, which Retains the memory, which will result in less pinning in the case
// where the underlying memory is backed by pre-pinned buffers.
return MemoryMarshal.TryGetArray(memory, out ArraySegment<byte> arraySegment) && ReferenceEquals(buffer, stream._buffer) ?
    new FileStreamCompletionSource(stream, numBufferedBytesRead, buffer) :
    new MemoryFileStreamCompletionSource(stream, numBufferedBytesRead, memory);
```

Reply:
This is clearer.